[RFC] treewide: some desktop file cleanups

K900 opened the pull request:
###### Motivation for this change
Started off trying to fix desktop files for nixpkgs-mozilla Firefoxen, now we here.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

K900 committed:
build-support/makeDesktopItem: make fully declarative, add all missing options

This updates makeDesktopItem to explicitly support all the fields in the spec,
converts list-like fields to native Nix lists instead of semicolon-separated strings,
and allows automatically generating [Desktop Action] sections from Nix code
instead of hardcoding them as extraConfig strings.

K900 committed:
firefox/wrapper: allow specifying StartupWMClass

Primarily useful for out of tree beta/nightly builds, which use a different WM_CLASS

mweinelt commented:
Firefox changes are looking good to me.

K900 commented:
Ran nixpkgs-review, things built fine.

sersorrel commented on pkgs/applications/version-management/smartgithg/default.nix:
```diff
@@ -76,12 +76,10 @@ stdenv.mkDerivation rec {
-    '';
+    startupWMClass = name;
+    extraDesktopEntries = {
+      Keywords= "git";
```
```suggestion
      Keywords = "git";
```

K900 replied:
Thanks, applied.

K900 committed:
treewide: switch all desktop file generators to new API

Notably:
- remove explicit arguments that match the defaults
- convert everything to the right Nix types

K900 commented:
![image](https://user-images.githubusercontent.com/386765/155180930-69f43a53-2b7d-437b-8585-5b9633f02050.png)

danth commented on pkgs/build-support/make-desktopitem/default.nix:
```diff
@@ -1,67 +1,96 @@
+  renderLine = name: value: if value != null then "${name}=${value}" else null;
+
+  renderSection = sectionName: attrs:
+    let
+      lines = builtins.filter (v: v != null) (lib.mapAttrsToList renderLine attrs);
+    in
+      builtins.concatStringsSep "\n" ([ "[${sectionName}]" ] ++ lines);
```
IMO use of `pipe` here would be easier to understand:
```suggestion
    lib.pipe attrs [
      (lib.mapAttrsToList renderLine)
      (builtins.filter (v: !isNull v))
      (builtins.concatStringsSep "\n")
      (section: ''
        [${sectionName}]
        ${section}
      '')
    ];
```

K900 replied:
Thanks, I do like this more. Applied.

danth reviewed (approved):
Looked through all the changes, nothing seems out of place to me.

jonringer reviewed (approved):
did a very quick scan, but everything looks good.

K900 commented:
Pushed a small fix to make desktop files with actions generate correctly again.

piegamesde reviewed (commented):
Change generally looks good, but I have a few nits:

- This does not appear to be backwards compatible, so it should be posted in https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574
- Please add a few code comments explaining the most critical sections, what is happening and why.
- Same thing would be cool for the commit message as well

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/7

K900 committed:
nixos/doc: add release note for makeDesktopItem changes

K900 commented:
Added some comments and wrote up a release note.

jonringer commented:
I can see this getting stale pretty quickly. are we good to merge?

K900 commented:
I believe so, but let me run one last rebase + eval.

mweinelt commented:
```
these 2 derivations will be built:
  /nix/store/1iiakhv3l2cchhbf9z6s64jpqm5wad1h-PrusaSlicer-G-code-Viewer.desktop.drv
  /nix/store/b53py1ghfr4infy3w74lckcqqcsqk0rm-prusa-slicer-2.4.0.drv
building '/nix/store/1iiakhv3l2cchhbf9z6s64jpqm5wad1h-PrusaSlicer-G-code-Viewer.desktop.drv'...
/build/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 1: G-code: command not found
error: builder for '/nix/store/1iiakhv3l2cchhbf9z6s64jpqm5wad1h-PrusaSlicer-G-code-Viewer.desktop.drv' failed with exit code 127
error: 1 dependencies of derivation '/nix/store/b53py1ghfr4infy3w74lckcqqcsqk0rm-prusa-slicer-2.4.0.drv' failed to build
```

I wonder if this change somehow broke `prusa-slicer`.

K900 commented:
I think I know what happened, I'll prepare a fix shortly.

sternenseemann commented:
This broke `freeoffice`.

K900 commented:
I did evaluate it, but it looks like something went wrong during the final rebase. Fix is at https://github.com/NixOS/nixpkgs/pull/162488

oxalica commented:
I noticed that there are still many manual `cp` or `ln` used with `makeDesktopItem`, while we already have the `copyDesktopItems` hook. Should we do another clean up?

K900 commented:
Should be a good idea, but I don't think I have the brain space to go through all 300 of these again :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/meaning-of-error-via-home-manager-error-makedesktopitem-value-for-categories-must-be-a-list/18648/2
